pkg/manage/cmd_tunnel.go

```diff
@@ -0,0 +1,271 @@
+			*addrsF = append(*addrsF, a)
+		}
+		addrs := tunnelParseArgument(*addrsF)
+
+		var wg sync.WaitGroup
+		for local, remote := range addrs {
+			wg.Add(1)
+			go func(local, remote string) {
+				defer wg.Done()
+				if err := newTunnel(service, local, remote); err != nil {
+					log.Printf("Error connecting %s to %s: %v", local, remote, err)
+					return
+				}
+			}(local, remote)
+		}
+
+		wg.Wait()
+
+		return nil
+	}
+	return &cmd
+}
+
+func tunnelParseArgument(args []string) map[string]string {
+	m := make(map[string]string, len(args))
+	for _, arg := range args {
+		parts := strings.Split(arg, ":")
+		if len(parts) == 3 {
+			parts = append([]string{""}, parts...)
+		}
+		m[parts[0]+":"+parts[1]] = parts[2] + ":" + parts[3]
+	}
+	return m
+}
+
+// newTunnel creates a new tunnel via grpc to the manage service.
+//
+// Listens on the given localAddr, sends all data via grpc to the manage server
+// and there redirect it to the remoteAddr.
+//
+// Blocks until the tunnel is closed.
+func newTunnel(service proto.ManageClient, localAddr string, remoteAddr string) error {
+	// Listen on localAddr
+	lst, err := net.Listen("tcp", localAddr)
+	if err != nil {
+		return fmt.Errorf("start listening on %s: %v", localAddr, err)
+	}
+	defer lst.Close()
+	log.Printf("Listen on %s", localAddr)
+
+	// Waiting for connections
+	for {
+		conn, err := lst.Accept()
+		if err != nil {
+			log.Printf("Error accepting connection: %v", err)
+			continue
+		}
+
+		go func(conn net.Conn) {
+			defer conn.Close()
+
+			// Open Tunnel
+			ctx := metadata.NewOutgoingContext(
+				context.Background(),
+				metadata.Pairs("addr", remoteAddr),
+			)
+			tunnel, err := service.Tunnel(ctx)
+			if err != nil {
+				log.Printf("Error creating tunnel: %v", err)
+				return
+			}
+
+			// Connecting the local connection to the tunnel
+			if err := copyStream(tunnel, conn); err != nil {
+				log.Printf("Error tunneling data: %v", err)
+				return
+			}
+		}(conn)
+	}
+}
+
+// Tunnel redirects a package to a different service.
+func (s *Server) Tunnel(ts proto.Manage_TunnelServer) error {
+	md, ok := metadata.FromIncomingContext(ts.Context())
+	if !ok {
+		return fmt.Errorf("unable to get metadata from context")
+	}
+	addr := md.Get("addr")
+	if len(addr) != 1 {
+		return fmt.Errorf("expect one address (host:port) in the meta data")
+	}
+
+	conn, err := new(net.Dialer).DialContext(ts.Context(), "tcp", addr[0])
+	if err != nil {
+		return fmt.Errorf("connecting to %s: %v", addr[0], err)
+	}
+	defer conn.Close()
+
+	if err := copyStream(ts, conn); err != nil {
+		return fmt.Errorf("connection grpc to server: %v", err)
+	}
+
+	return nil
+}
+
+// SendReceiver reads and writes from a grpc tunnel connection.
+type SendReceiver interface {
+	Recv() (*proto.TunnelData, error)
+	Send(*proto.TunnelData) error
+}
+
+// copyStream connects the grcp connection with a io.ReadWriter.
+//
+// Blocks until one connection is closed.
+func copyStream(sr SendReceiver, rw io.ReadWriter) error {
```

Comment:
Maybe split this function up into three. One "main" copy function and two functions for the goroutines.

Reply:
This was a good idea. Done.